langroid/agent/chat_agent.py

```diff
@@ -623,6 +799,139 @@ def disable_message_use_except(self, message_class: Type[ToolMessage]) -> None:
             self.llm_tools_usable.discard(r)
             self.llm_functions_usable.discard(r)
 
+    def _load_output_format(self, message: ChatDocument) -> None:
+        """
+        If set, attempts to parse a value of type `self.output_format` from the message
+        contents and assigns it to `content_any`.
+        """
+        if self.output_format is not None:
+            try:
+                content_any = self.output_format.parse_obj(json.loads(message.content))
+
+                if issubclass(self.output_format, PydanticWrapper):
+                    message.content_any = content_any.value  # type: ignore
+                else:
+                    message.content_any = content_any
+            except ValidationError:
+                self.disable_strict = True
+                logging.warning(
+                    """
+                    Validation error occured with strict output format enabled.
+                    Disabling strict mode.
+                    """
+                )
+            except json.JSONDecodeError:
+                self.disable_strict = True
+                logging.warning(
+                    """
+                    JSON decode error occured with strict output format enabled.
+                    Disabling strict mode.
+                    """
+                )
+
+    def get_tool_messages(
+        self,
+        msg: str | ChatDocument | None,
+        all_tools: bool = False,
+    ) -> List[ToolMessage]:
+        """
+        Extracts messages and tracks whether any errors occured. If strict mode
+        was enabled, disables it for the tool, else triggers strict recovery.
+        """
+        self.tool_error = False
+        try:
+            tools = super().get_tool_messages(msg, all_tools)
+        except ValidationError as ve:
+            tool_class = ve.model
+            if issubclass(tool_class, ToolMessage):
+                was_strict = (
+                    self.config.use_functions_api
+                    and self.config.use_tools_api
+                    and self._strict_mode_for_tool(tool_class)
+                )
+                if was_strict:
+                    name = tool_class.default_value("request")
+                    self.disable_strict_tools_set.add(name)
+                    logging.warning(
+                        f"""
+                        Validation error occured with strict tool format.
+                        Disabling strict mode for the {name} tool.
+                        """
+                    )
+                else:
+                    self.tool_error = True
```

Comment:
short comment to explain this, e.g. what does `tool_error` signify?

Reply:
Added comments; in strict mode this means there was some schema (or perhaps token limit) error and we should disable, but in non-strict mode we can trigger strict recovery and correct.